### PR TITLE
Add additional network healthchecks

### PR DESCRIPTION
**Pull request by avilevy18**
## Related issue
b/228050119

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [x] Integration tests do not apply.
  - [ ] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

File: internal/healthchecks/error.go
Comment (rafaelwestphal) [inline internal/healthchecks/error.go:98]: Maybe 
https://cloud.google.com/logging/docs/agent/ops-agent/troubleshoot-find-info
is a better landing page for these connectivity troubleshooting errors.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (rafaelwestphal) [inline cmd/google_cloud_ops_agent_engine/main.go:33]: "only run health checks" or "run health checks and exit". This can be a bit ambiguous as to "We are enabling/disabling health checks when running the service", so a different wording might be better.

File: internal/healthchecks/network_check.go
Comment (rafaelwestphal) [inline internal/healthchecks/network_check.go:38]: Although I don't mind that much adding these urls to the health checks, they are more valuable at the diagnostics script and don't add much value after the agent is already installed. Since they refer to the endpoints that will enable the installation to succeed, without access to packages and/or dl.google you can't download the Ops-Agent and run the health checks.

**Comment by rafaelwestphal** [comment]
Thanks! I'll give @franciscovalentecastro a chance to look at these changes, since he is working closely on this.

File: internal/healthchecks/network_check.go
Comment (franciscovalentecastro) [reply internal/healthchecks/network_check.go]: This is a great catch that I haven't realized before. It doesn't hurt to check them here, but there is more value in the diagnostics script if the user can't even install the Ops Agent.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (franciscovalentecastro) [inline cmd/google_cloud_ops_agent_engine/main.go:56]: I have the concern that running with `--healthchecks` flag is going to run all of the Health Checks (Ports, Network, API), but the `Ports Check` is going to fail if the Ops Agent is running, causing confusion in the results. 

We can disable that `Ports Check` (not registering that check for this case) to avoid this.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (franciscovalentecastro) [inline cmd/google_cloud_ops_agent_engine/main.go:73]: Is there a way to only have one entry point for the `runHealthChecks` that works for either the Ops Agent startup or just the `--healthchecks` flag.

File: internal/healthchecks/network_check.go
Comment (franciscovalentecastro) [reply internal/healthchecks/network_check.go]: Checking the `metadata` server here still has a lot of value.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (jefferbrecht) [reply cmd/google_cloud_ops_agent_engine/main.go]: What if we just make the ports check short-circuit as a `PASS` if the Ops Agent (and subagents) are already running?

Comment (franciscovalentecastro) [reply cmd/google_cloud_ops_agent_engine/main.go]: Yeah, that's a good idea. I like it better than disabling the `Ports Check`. What is the best way to detect the Ops Agent is already running ?

File: internal/healthchecks/network_check.go
Comment (jefferbrecht) [inline internal/healthchecks/network_check.go:23]: Why not just use a slice here?

File: internal/healthchecks/error.go
Comment (jefferbrecht) [inline internal/healthchecks/error.go:59]: Is any of this used?

File: internal/healthchecks/error.go
Comment (jefferbrecht) [inline internal/healthchecks/error.go:112]: Message looks incomplete

File: internal/healthchecks/network_check.go
Comment (jefferbrecht) [inline internal/healthchecks/network_check.go:23]: Optional: since `requests` is only used once, you could define both the struct and its elements inline within `RunCheck` and avoid needing to define a named struct. (This is a common pattern used in unit tests too: https://gobyexample.com/testing)

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (jefferbrecht) [reply cmd/google_cloud_ops_agent_engine/main.go]: I guess there are at least two ways:

1. Check that the subagent services are running. If they're successfully running then necessarily they must be holding onto the ports, so it's a pass.
2. Check which processes (if any) own ports 20201 and 20202; if they match what we expect (e.g. `/opt/whatever/fluent-bit` and `/opt/whatever/otelopscol` for Linux, likewise for Windows) then the subagents are already running with those ports, so it's a pass.

I am probably missing some gotchas here so I'm curious to know what others think.

Comment (igorpeshansky) [reply cmd/google_cloud_ops_agent_engine/main.go]: 2 might be somewhat problematic, since checking which process owns a port is often a privileged operation, and the health check invocation will likely be in user space… 1 could be sufficient.

Comment (franciscovalentecastro) [reply cmd/google_cloud_ops_agent_engine/main.go]: Yes, I'm also curious about what others think. This came up at the last discussion relating the `Ports Check Flakiness`. 

I would go with `1.` (but still not sure how to do it). `2.` (as mentioned by @igorpeshansky) would be problematic since probably running a `netstat` command or something similar is going to be platform/implementation dependent.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (franciscovalentecastro) [inline cmd/google_cloud_ops_agent_engine/main.go:36]: Also, we should also add this in the Windows service too :
https://github.com/GoogleCloudPlatform/ops-agent/blob/master/cmd/ops_agent_windows/run_windows.go#L136-L151

If this feels to big of a change, maybe we can only add the new `Network Checks` here and defer the `--healthchecks` fucntion to a followup `PR`.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (jefferbrecht) [reply cmd/google_cloud_ops_agent_engine/main.go]: On Linux: check the exit code of `systemctl is-active --quiet google-cloud-ops-agent-fluent-bit` (likewise for the metrics service)

On Windows: check the output of `sc query google-cloud-ops-agent-fluent-bit | findstr "STATE"` for `RUNNING` (likewise for the metrics service)

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (franciscovalentecastro) [reply cmd/google_cloud_ops_agent_engine/main.go:66]: I think this was addressed.

File: cmd/google_cloud_ops_agent_engine/main.go
Comment (franciscovalentecastro) [inline cmd/google_cloud_ops_agent_engine/main.go:68]: Some suggestions for this : 
- Moving the execution of the `healthchecks` to the beginning of `run()` so it's the first thing. The reasoning for this is if `--healthchecks` is set, then it will only run the `healthchecks` and no other part of the confgenerator.
- I think there's no need anymore to pass as a paramter `healthchecks.HealthCheckRegistryFactory()`.
- Maybe add a comment to the `return nil` explaining why it's set this way.

File: internal/healthchecks/linux.go
Comment (franciscovalentecastro) [inline internal/healthchecks/linux.go:30]: What happens in the case the `service exists`, but it is not running.

File: cmd/ops_agent_windows/main_windows.go
Comment (franciscovalentecastro) [inline cmd/ops_agent_windows/main_windows.go:67]: Should we use `infoLog` here ?

File: internal/healthchecks/ports_check.go
Comment (franciscovalentecastro) [inline internal/healthchecks/ports_check.go:52]: I think the name of the `opentelemetry-collector` subagent in windows is different to the one in linux. See https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/ops_agent_test.go#L144-L150

Comment (jefferbrecht) [reply internal/healthchecks/ports_check.go:52]: That's the process name -- we're checking the service name here, which looks correct: https://github.com/GoogleCloudPlatform/ops-agent/blob/master/cmd/ops_agent_windows/main_windows.go#L120

File: internal/healthchecks/linux.go
Comment (avilevy18) [reply internal/healthchecks/linux.go:30]: The command exits with status code 3 so the code will run as if the service doesn't exists.

File: internal/healthchecks/ports_check.go
Comment (franciscovalentecastro) [reply internal/healthchecks/ports_check.go:52]: Ok! Thanks for the clarification.

**Review by franciscovalentecastro — APPROVED** [review]
LGTM! Approved modulo addressing the last comments regarding refactors.

File: internal/healthchecks/ports_check.go
Comment (jefferbrecht) [reply internal/healthchecks/ports_check.go:48]: I don't think we should merge them, because there exist intermediate states (especially on Windows) where only one subagent is running. Each subagent uses a different port so the checks divide pretty cleanly anyway.

File: cmd/ops_agent_windows/main_windows.go
Comment (jefferbrecht) [inline cmd/ops_agent_windows/main_windows.go:65]: Does `result.Err` contain anything relevant that we might want to include in the printed text?

File: internal/healthchecks/network_check.go
Comment (jefferbrecht) [inline internal/healthchecks/network_check.go:69]: Currently `RunCheck` will fail fast on the first encountered failure. Would it make more sense to run all of the checks independently and aggregate the result with `multierr` instead of failing fast?

File: internal/healthchecks/linux.go
Comment (jefferbrecht) [inline internal/healthchecks/linux.go:30]: Exit code 3 means "unit is not active", not "service doesn't exist".

File: internal/healthchecks/network_check.go
Comment (franciscovalentecastro) [reply internal/healthchecks/network_check.go:69]: Yes, it should report an aggregate result for the failures, not just fail fast. We (@avilevy18, @braydonk) discussed this at beginning of this task. 

I will address this in the followup PR https://github.com/GoogleCloudPlatform/ops-agent/pull/1139 , since the `API Check` (and now it seems also the `Ports Check`) will benefit from reporting all of the fails/errors. Initially (the current state of #1139) I was only going to separate in different checks (e.g. Log API and Mon API checks), but after discussing the solution was to create a `type MultiHealthCheckError []HealthCheckError` and report each of the relevant failures.

File: internal/healthchecks/network_check.go
Comment (jefferbrecht) [inline internal/healthchecks/network_check.go:59]: Would it make sense (maybe as a later refactor after this PR) to merge this struct with the one from `error.go`? Everything looks like it's one-to-one so it might make more sense to just have everything related to a check be all in one place...

Comment (avilevy18) [reply internal/healthchecks/network_check.go:59]: Lets refactor in another another PR.

File: cmd/ops_agent_windows/main_windows.go
Comment (jefferbrecht) [reply cmd/ops_agent_windows/main_windows.go:65]: Discussed offline; `result.Message` already embeds the contents of `result.Err`.

I would suggest a small refactor of `HealthCheckResult`:
* Get rid of the `Message` field and change it into a `String()` function that generates the formatted message on-demand. This makes it so that we don't have to manually keep `Err` and `Message` in sync, with the added benefit of implementing `Stringer`.
* Add a `Name` field to support the above function since it's needed for generating the message.

WDYT?

Comment (franciscovalentecastro) [reply cmd/ops_agent_windows/main_windows.go:65]: I really like this idea of generating the formatted message on demand. I was trying to figure out where to put this and `result` seems to be great (it has all the info, with the `Err` and name it should be enough).

File: internal/healthchecks/error.go
Comment (avilevy18) [reply internal/healthchecks/error.go:95]: Spoke offline, more robust tests will be available in https://github.com/GoogleCloudPlatform/ops-agent/pull/1139

File: cmd/ops_agent_windows/main_windows.go
Comment (avilevy18) [reply cmd/ops_agent_windows/main_windows.go:65]: Spoke offline, will be part of https://github.com/GoogleCloudPlatform/ops-agent/pull/1139

File: cmd/ops_agent_windows/main_windows.go
Comment (jefferbrecht) [inline cmd/ops_agent_windows/main_windows.go:69]:
```suggestion
			fmt.Println("Use either --healthchecks, --install, --uninstall to take action.")
```

File: internal/healthchecks/ports_check.go
Comment (jefferbrecht) [inline internal/healthchecks/ports_check.go:90]: Nit:

```suggestion
	// Opentelemetry-collector listens in both tcp4 and tcp6. Check for opentelemetry-collector self metrics port.
```

**Review by jefferbrecht — APPROVED** [review]
LGTM, just a couple typos remaining.

File: internal/healthchecks/healthchecks.go
Comment (franciscovalentecastro) [inline internal/healthchecks/healthchecks.go:88]: There is no need to add `Error: ` as prefix here, since the message already will be formatted as, e.g.`API Check - Result: ERROR, Detail: ****`

**Review by franciscovalentecastro — APPROVED** [review]

**Comment by jefferbrecht** [comment]
If the helper refactor is causing issues then feel free to revert it. What I had in mind was something like this:

```go
func LogHealthCheckResults(healthCheckResults map[string]HealthCheckResult, infoLogger func(string), errorLogger func(string)) {
	for _, result := range healthCheckResults {
		if result.Err != nil {
			errorLogger(result.Message)
		} else {
			infoLogger(result.Message)
		}
	}
}
```

Which can be invoked e.g. on Linux as:

```go
LogHealthCheckResults(healthCheckResults, func(s string) { logger.Println(s) }, func(s string) { logger.Println(s) })
```

Or on Windows as:

```go
LogHealthCheckResults(healthCheckResults, func(s string) { s.Log.Info(EngineEventID, s) }, func(s string) { s.Log.Error(EngineEventID, s) })
```

Or if you need stdout:

```go
LogHealthCheckResults(healthCheckResults, func(s string) { log.Println(s) }, func(s string) { log.Println(s) })
```

**Pull request merged**
